Usbc source

sambhurst opened the pull request:
Adds Power Delivery Source Support to the USB-C Stack. See [Issue 38371](https://github.com/zephyrproject-rtos/zephyr/issues/38371) for details

sambhurst committed:
samples: usb-c: source: Add USB-C Subsystem Source Sample

Implementing USB-C Source functionality can be difficult.
This sample application serves as an example of how
to create an application with Power Delivery Source
functionality.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
doc: Document the USB-C Source Subsystem

Documents the USB-C Source Subsystem API.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

fabiobaltieri commented:
Hey @sambhurst, do you think it would make sense to regroup the commints in smaller PRs? Looks like there's a mix of refactoring/preparation, board specific changes, bug fixes and then the main feature. Thinking having more small PRs may make it easier to review and hopefully get it merged faster, and also make it easier to maintain the stack through the reviews.

sambhurst commented:
> Hey @sambhurst, do you think it would make sense to regroup the commints in smaller PRs? Looks like there's a mix of refactoring/preparation, board specific changes, bug fixes and then the main feature. Thinking having more small PRs may make it easier to review and hopefully get it merged faster, and also make it easier to maintain the stack through the reviews.

@fabiobaltieri I agree and have currently split out the refactoring to a separate PR[53071](https://github.com/zephyrproject-rtos/zephyr/pull/53071). I'll split out the others until there's just the Source PD stuff in this PR

sambhurst commented:
I've split out 9 files into 3 separate PRs
1: usb-c: Refactor USB-C Subsystem Sink [#53071](https://github.com/zephyrproject-rtos/zephyr/pull/53071)
2: usbc boards [#53073](https://github.com/zephyrproject-rtos/zephyr/pull/53073)
3: usbc tcpc stm32 [#53077](https://github.com/zephyrproject-rtos/zephyr/pull/53077)

alevkoy commented:
You could probably also split VCONN support and the GoodCRC timer into separate PRs. Those seem pretty orthogonal to sourcing.

Hypothetically, if you did that, would it just be the last 4 commits in this PR that still need to be reviewed?

DianaZig commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+	LOG_INF("UnattachedWait.SRC");
+
+	/* Start discharging VCONN */
+	tcpc_vconn_discharge(tcpc, true);
```
Does this call take care of ensuring Rp isn't on the pin being discharged?

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,755 @@
+	 */
+	else if ((atomic_test_bit(&pe->flags, PE_FLAGS_HAS_BEEN_PD_CONNECTED) == false)
+			&& usbc_timer_expired(&pe->pd_t_no_response)
+			&& pe->hard_reset_counter > PD_N_CAPS_COUNT) {
```
Is this the right counter comparison?

sambhurst replied:
Fixed

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,755 @@
+		if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+			union pd_header header = prl_rx->emsg.header;
+
+			if ((header.extended == false) &&
```
Extended check in the helper already?

sambhurst replied:
Fixed

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,755 @@
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	usbc_timer_start(&pe->pd_t_sender_response);
```
Timer stop?

sambhurst replied:
Code has been refactored

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,755 @@
+	 * PE_SRC_Transition_Supply state when:
+	 *	1) The Request can be met.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
```
Could the run state here just be collapsed into the entry state?

sambhurst replied:
Done

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,755 @@
+	 *	   the power supply is ready.
+	 */
+	if (atomic_test_bit(&pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		if (pe->snk_request_reply == SNK_REQUEST_VALID) {
```
How do we transition out if we sent GotoMin?  (alternatively, we might leave out GotoMin for the first round on the source...)

sambhurst replied:
Good catch. I've updated the code.

DianaZig reviewed (commented):
I've just looked over the source TC and half the source PE since I know the other files will probably be affected by the other CLs in flight

laurenmurphyx64 commented:
Hi @sambhurst, do you need to merge this PR or any other PRs to finish https://github.com/zephyrproject-rtos/zephyr/issues/38371 for the release? Asking because feature freeze is next week on Friday

sambhurst commented:
> Hi @sambhurst, do you need to merge this PR or any other PRs to finish #38371 for the release? Asking because feature freeze is next week on Friday

Hello @laurenmurphyx64, yes there are quite a few PRs that need to merge before #38371 is finished and unfortunately, it won't happen before next Friday.

teburd commented:
@sambhurst please rebase
@galak suggested thinking about splitting this up into smaller PRs for review

galak commented:
For example the commit - ' boards: arm: stm32g081b_eval: Add ADC1 Channel9 ' could get pulled out and probably merged.

sambhurst committed:
usb_c: Transmit Error are not Protocol Errors

Transmit errors must be treated differently than Protocol Errors.
This change sets a flag that informs the stack of a Message
Transmit error.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented:
@alevkoy @keith-zephyr please review when you have time.

alevkoy reviewed (commented):
First pass: Just looked through the headers. So far, non-blocking comments only.

sambhurst commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -294,6 +337,82 @@ void usbc_set_policy_cb_get_rdo(const struct device *dev,
+ * @param dev Runtime device structure
+ * @param cb callback
+ */
+void usbc_set_policy_cb_set_snk_cap(const struct device *dev,
```
I agree. I've added "port_partner" to the callback and function. I'll update the source caps callback in another PR

sambhurst commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -294,6 +337,82 @@ void usbc_set_policy_cb_get_rdo(const struct device *dev,
+					  const policy_cb_check_if_ps_ready_t cb);
+
+/**
+ * @brief Set the callback to check if Present Contract is still valid
```
Done

sambhurst commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -145,6 +157,18 @@ enum usbc_policy_wait_t {
+/**
+ * @brief Device Policy Manager's response to a Sink Request
+ */
+enum usbc_snk_req_reply_t {
```
I agree that this should be more general.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -96,13 +97,17 @@ static void pe_init(const struct device *dev)
 
 	/* Initialize common timers */
-	usbc_timer_init(&pe->pd_t_sender_response, PD_T_SENDER_RESPONSE_NOM_MS);
+	usbc_timer_init(&pe->pd_t_sender_response, PD_T_NO_RESPONSE_MAX_MS);
```
The sender response is handled in a parent state that's common to both source and sink.

alevkoy replied:
You're right. I'm not sure what I was thinking here.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -96,13 +97,17 @@ static void pe_init(const struct device *dev)
 	/* Initialize common counters */
 	pe->hard_reset_counter = 0;
 
+#ifdef CONFIG_USBC_CSM_SINK_ONLY
```
Probably and I've not put much thought into it yet.

sambhurst commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -48,6 +48,11 @@ config USBC_CSM_SINK_ONLY
 	  Allows the USB-C state machine to function as a Sink
 
+config USBC_CSM_SOURCE_ONLY
+	bool "Source USBC Conneciton State Machine"
```
Fixed

alevkoy commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+#include "usbc_tc_src_states_internal.h"
+
+/**
+ * @brief 4.5.2.2.7 Unattached.SRC State
```
Nit: I'd cite the spec version in case they change the section numbering later.

sambhurst replied:
Done

alevkoy commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+	/* Start sourcing VBUS */
+	if (data->policy_cb_src_en(dev, true) == 0) {
+		/* Start sourcing VCONN */
+		if (policy_check(dev, CHECK_VCONN_CONTROL)) {
```
Should VCONN support be its own config, or do you want to leave it entirely to policy?

sambhurst replied:
Let's leave it as policy for now. We can add it to the config in a future PR if we need to.

alevkoy commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+ *
+ *  2: The port shall continue to provide an Rp termination on the CC pin not
+ *     being discharged.
+ *     NOTE: Implemented in last TC source state.
```
I thought this was controlled by setting Rp on entering the Rp super state.

sambhurst replied:
I've updated the comment

alevkoy commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+	tcpc_vconn_discharge(tcpc, true);
+
+	/* Use debounce timer to give VCONN time to discharge */
+	usbc_timer_start(&tc->tc_t_cc_debounce);
```
Nit: The nominal limit for this process is tVconnOff (Type-C table 4-31). I would guess that the actual limit is hardware dependent. We might want a way for the TCPC driver or DPM to influence this.

sambhurst replied:
I agree but for now, I've changed the code to use the proper timer. Let's add your suggested change in a future PR.

alevkoy reviewed (commented):
Another pass. Looked at the TC changes.

sambhurst commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+		new_cc_state = TC_CC_UFP_ATTACHED;
+	} else {
+		/* No UFP */
+		tc_set_state(dev, TC_UNATTACHED_SRC_STATE);
```
Yes.

sambhurst commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+
+	/* Wait for CC debounce */
+	if (usbc_timer_running(&tc->tc_t_cc_debounce) &&
+		usbc_timer_expired(&tc->tc_t_cc_debounce) == false) {
```
I've made the change

sambhurst commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,333 @@
+	 * and the SRC.Rd state is detected on exactly one of the CC1 or CC2
+	 * pins for at least tCCDebounce.
+	 *
+	 * If the port supports Debug Accessory Mode, it shall transition to
```
I've removed the comment

alevkoy commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -755,13 +805,63 @@ static void prl_tx_snk_start_ams_run(void *obj)
+		atomic_clear_bit(&prl_tx->flags, PRL_FLAGS_MSG_XMIT);
+
+		/*
+		 * Soft Reset Message pending &
```
Nit: Weird line break here and below

sambhurst replied:
fixed

alevkoy commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -1164,6 +1264,18 @@ static const struct smf_state prl_tx_states[PRL_TX_STATE_COUNT] = {
 		prl_tx_suspend_run,
 		NULL,
 		NULL),
+#ifdef CONFIG_USBC_CSM_SOURCE_ONLY
```
Should the sink states also be config-guarded?

sambhurst replied:
done

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -353,6 +358,17 @@ bool pe_dpm_initiated_ams(const struct device *dev)
+/**
+ * @brief End and atomic messaging sequence
+ */
+void pe_dpm_end_ams(const struct device *dev)
```
Right now, you set this flag directly and sometimes clear it directly. It would be nice to handle consistently via these accessors.

sambhurst replied:
I've made the change for clearing the bit. The other accessors aren't used in this file.

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -584,6 +602,83 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
+}
+
+/**
+ * @brief Check if the Power Supply is ready
```
Nit: Why capitalize power supply?

sambhurst replied:
fixed

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -716,6 +813,86 @@ static void pe_drs_send_swap_run(void *obj)
+				} else if (received_control_message(dev, header, PD_CTRL_REJECT) ||
+					received_control_message(dev,
+							header, PD_CTRL_NOT_SUPPORTED)) {
+#ifdef CONFIG_USBC_CSM_SOURCE_ONLY
```
Upstream would probably benefit from a `pe_set_ready_state` function like we have downstream.

sambhurst replied:
I agree and this will be added with the DRP PR

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -716,6 +813,86 @@ static void pe_drs_send_swap_run(void *obj)
+					pe_set_state(dev, PE_SNK_READY);
+#endif
+				}
+				/* Unexpected messages fall through to soft reset */
```
Without some `return`s above, all messages will fall through to soft reset.

sambhurst replied:
good catch. fixed

alevkoy commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -941,14 +1118,52 @@ static void pe_sender_response_run(void *obj)
-
+#else
+		/* Source states */
+		case PE_SRC_DISCOVERY:
```
Instead of putting these state-specific things in this super state, which will need to get bigger and more complicated over time, what if the super state set a flag when the timer expired, and each child state checked for it?

sambhurst replied:
This would add delay because the state machine will need to cycle again in order for the flag to be detected.

alevkoy reviewed (commented):
Looked over the PRL (not my area of expertise) and started on the PE.

alevkoy commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+			case PD_CTRL_GET_SOURCE_CAP:
+				pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+				break;
+				break;
```
Extra break

sambhurst replied:
fixed

alevkoy commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+			 * Receiving an unexpected message shall be responded
+			 * to with a soft reset message.
+			 */
+			case PD_CTRL_ACCEPT:
```
Practically, I think it's OK to ignore unexpected GoodCRC and Not_Supported, but does the spec say that we should treat these differently than an unexpected Reject or Accept? Should an unexpected GoodCRC even make it past the PRL to the PE?

sambhurst replied:
Both messages are passed from the PRL to the PE and are expected in this state. Good_Crc should be ignored but it's probably a policy decision as to what to do when Not_Supported is received.

alevkoy commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+/**
+ * @brief 8.3.3.2.9 PE_SRC_Hard_Reset State
+ */
+void pe_src_hard_reset_entry(void *obj)
```
It seems like PE_SRC_Hard_Reset_Received and PE_SRC_Hard_Reset could share most of their code.

sambhurst replied:
I moved common code to a parent state

alevkoy replied:
It looks like you made `PE_SRC_HARD_RESET_RECEIVED` its own parent with no implementation. Should it have `PE_SRC_HARD_RESET_PARENT` as its parent?

sambhurst replied:
Good catch.

alevkoy commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	 *	1: request the Device Policy Manager to turn on VCONN
+	 *	2: inform the Protocol Layer that the Hard Reset is complete.
+	 *
+	 * NOTE: The Device Policy Manager turns on VCONN when it send the
```
"The Device Policy Manager turns on VCONN when it notifies the PE that Power Supply is at the default level"?

sambhurst replied:
fixed

alevkoy reviewed (commented):
Another pass. Made it through the implementation.

alevkoy commented on doc/connectivity/usb_c/ucds.rst:
```diff
@@ -108,6 +108,96 @@ Start the USB-C subsystem:
+
+The following devicetree, structures and callbacks need to be defined:
+
+* Devicetree usb-c-connector node referencing a TCPC
```
Here and below, `usb-c-connector` and `vbus` should probably have double backticks or some other markup to indicate that they are the names of DTS nodes.

sambhurst replied:
done

alevkoy reviewed (commented):
I think I made a complete pass.

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -294,6 +337,82 @@ void usbc_set_policy_cb_get_rdo(const struct device *dev,
+ * @param dev Runtime device structure
+ * @param cb callback
+ */
+void usbc_set_policy_cb_get_src_rp(const struct device *dev,
+				   const policy_cb_get_src_rp_t cb);
```
General comment regarding this expansion of the API.  Would it make sense to group these policy callbacks by function? Something like:
```
void usbc_set_policy_cb();  /* Shared callbacks needed for src and snk */
void usbc_set_policy_cb_snk(); /* snk only callbacks */
void usbc_set_policy_cb_src(); /* src only callbacks */
```

sambhurst replied:
Will make change in a separate PR.

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	 *	1) The Device Policy Manager informs the Policy Engine that
+	 *	   the power supply is ready.
+	 */
+	if (atomic_test_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
```
When does TX_COMPLETE clear?

sambhurst replied:
See the pe_send_xxx_msg functions in usbc_pe_common.c

```
void pe_send_data_msg(const struct device *dev, const enum pd_packet_type type,
                        const enum pd_data_msg_type msg)
{
        struct usbc_port_data *data = dev->data;
        struct policy_engine *pe = data->pe;

        /* Clear any previous TX status before sending a new message */
        atomic_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE);
        prl_send_data_msg(dev, type, msg);
}
```

DianaZig commented on subsys/usb/usb_c/usbc_pe_common_internal.h:
```diff
@@ -302,6 +362,17 @@ bool policy_wait_notify(const struct device *dev, const enum usbc_policy_wait_t
+						const uint32_t request_msg);
+
+/**
+ * Check if the Present Contract is still valid
```
After having now re-read this part of the spec a few times, we should probably include more details here for developer reference.  The contract is considered "invalid" if the previous current/voltage are no longer available AND the sink fails to make a valid request.  

This seems like something the PE "could" take care of for the DPM, but I'm not sure if we want to or not.  The spec doesn't mention that the DPM has to be the arbiter of this determination, but the DPM is tracking the source caps generally.

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	struct policy_engine *pe = data->pe;
+
+	if (pe->dpm_request > REQUEST_TC_END) {
+		atomic_set_bit(pe->flags, PE_FLAGS_DPM_INITIATED_AMS);
```
It seems like this bit will be set even if we didn't change states to start a new AMS?

sambhurst commented on doc/connectivity/usb_c/ucds.rst:
```diff
@@ -108,6 +108,96 @@ Start the USB-C subsystem:
+
+The configuration of a USB-C Device is done in the stack layer and devicetree.
+
+The following devicetree, structures and callbacks need to be defined:
```
done

sambhurst commented on doc/connectivity/usb_c/ucds.rst:
```diff
@@ -14,7 +14,7 @@ It provides the following functionalities:
   :zephyr_file:`include/zephyr/usb_c/usbc.h`
 
-Currently the device stack only support implementation of Sink devices.
+Currently the device stack support implementation of Sink and Source devices.
```
done

sambhurst commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -294,6 +337,82 @@ void usbc_set_policy_cb_get_rdo(const struct device *dev,
+					   const policy_cb_check_sink_request_t cb);
+
+/**
+ * @brief Set the callback used to chedk if Source Power Supply is ready
```
The code was wrong. It has been fixed

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	 *	   the power supply is ready.
+	 */
+	if (atomic_test_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		if (policy_check_if_ps_ready(dev)) {
```
You're right and good catch. The code is implemented wrong. I should send a policy notify to start transitioning the PS and only use policy_check_if_ps_ready to only check if it's ready.

sambhurst commented on samples/subsys/usb_c/source/src/main.c:
```diff
@@ -0,0 +1,363 @@
+	struct port1_data_t *dpm_data = usbc_get_dpm_data(dev);
+
+	/* Transition Power Supply to new voltage */
+	source_ctrl_set(dpm_data->obj_pos);
```
good catch. I changed the code.

sambhurst commented on samples/subsys/usb_c/source/src/main.c:
```diff
@@ -0,0 +1,363 @@
+#define PORT1_POWER_ROLE	DT_ENUM_IDX(DT_NODELABEL(port1), power_role)
+
+/* A Source power role evauates to 1. See usbc_tc.h: TC_ROLE_SOURCE */
+#if (PORT1_POWER_ROLE != 1)
```
I tried that before changing it to 1. The enum isn't recognized during this pass.

alevkoy replied:
Ack.

sambhurst commented on samples/subsys/usb_c/source/README.rst:
```diff
@@ -0,0 +1,81 @@
+************
+The TCPC device used by the sample is specified in the devicetree
+node that's compatible with ``usb-c-connector``.
+The sample has been tested on :ref:`stm32g081b_eval_board`.
```
fixed

sambhurst commented on samples/subsys/usb_c/source/README.rst:
```diff
@@ -0,0 +1,81 @@
+using the USB-C subsystem. The application implements a USB-C Source device.
+
+After a USB-C Sink device is plugged into the USB-C Source device, it
+negotiates with the Source device to provide 5V@100mA, 9V@100mA, or 15V@100mA
```
done

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	const struct device *dev = pe->dev;
+
+	/* Send PS_RDY message */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
```
done

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	data->policy_cb_get_src_caps(dev, &pdos, &num_pdos);
+
+	msg->len = PD_CONVERT_PD_HEADER_COUNT_TO_BYTES(num_pdos);
+	memcpy(msg->data, (uint8_t *)pdos, msg->len);
```
yes. pdos is a uint32_t *

sambhurst replied:
Yeah, you are correct. I was thinking len was the number of objects not the number of bytes.

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+	if (usbc_timer_expired(&pe->pd_t_typec_send_source_cap)
+			&& pe->caps_counter <= PD_N_CAPS_COUNT) {
+		pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+	}
```
entering the disabled state when nCapsCount is reached is optional behavior. The nHardResetCount his handled in the super state.

alevkoy replied:
Ack. I left open the comment in the super state about how this could be easier to follow with different organization.

Do you plan to add a transition to PE_SRC_Disabled in a later PR or leave it as is?

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+		if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+			pe->hard_reset_counter = 0;
+			pe->caps_counter = 0;
+			pe->submachine++;
```
done

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,702 @@
+ * @brief USB-C Power Policy Engine (PE)
+ *
+ * The information in this file was taken from the USB PD
+ * Specification Revision 3.0, Version 2.0
```
Good question and it should probably be discussed in an issue. Could you open an issue and drive the discussion?

alevkoy replied:
I created https://github.com/zephyrproject-rtos/zephyr/issues/56697.

alevkoy replied:
I'm willing to leave this as-is for this PR, but let me know what you think on the issue. Also @keith-zephyr and @DianaZig.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -716,6 +813,86 @@ static void pe_drs_send_swap_run(void *obj)
+	switch (pe->submachine) {
+	case 0:
+		if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+			pe->submachine++;
```
fixed

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -716,6 +813,86 @@ static void pe_drs_send_swap_run(void *obj)
+	 */
+	pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_DR_SWAP);
+	/* Initialize Submachine */
+	pe->submachine = 0;
```
fixed

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -584,6 +602,83 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
+
+/**
+ * @brief Ask the DPM to change the Source Caps.
+ *	  Returns true if source Caps have been updated, else false
```
fixed

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -584,6 +602,83 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
+}
+
+/**
+ * @brief Check if the Present Contract is still valid
```
fixed

sambhurst commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -47,6 +47,8 @@ enum prl_flags {
 	 */
 	PRL_FLAGS_MSG_XMIT = 7,
+	/* Flag to note that PRL requested to set SINK_NG CC state */
+	PRL_FLAGS_SINK_NG			= 8,
```
fixed

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -264,13 +264,16 @@ void pe_report_error(const struct device *dev, const enum pe_error e,
 	/* All error types besides transmit errors are Protocol Errors. */
-	if ((e != ERR_XMIT && atomic_test_bit(pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS) == false) ||
-	    e == ERR_XMIT ||
+	if ((e != ERR_XMIT) && ((atomic_test_bit(pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS) == false) ||
```
Good point. I added your code plus some to make things work correctly.

keith-zephyr replied:
Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/52993/commits/cdc2e97f749fc203f231bc7db548a5aa0062ba52

sambhurst commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2022 The Chromium OS Authors
```
done

sambhurst commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -755,13 +805,63 @@ static void prl_tx_snk_start_ams_run(void *obj)
 	LOG_INF("PRL_Tx_Snk_Pending");
+
+	/* Stop SinkTxTimer */
+	usbc_timer_stop(&prl_tx->pd_t_sink_tx);
```
Absolutely. Not sure what I was thinking.

sambhurst commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -69,6 +69,16 @@ enum usbc_policy_request_t {
+	 * Request Policy Engine to request the port partner to source
+	 * minimum power
+	 */
+	REQUEST_PE_GOTO_MIN,
```
The code for GotoMin is in the stack and I left the "should" up the user's application.

DianaZig commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -229,6 +234,40 @@ void pe_message_sent(const struct device *dev)
+	/* Not in an explicit contract */
+	if (!atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT)) {
+		/* Protocol Errors occurred on SOP during an Interruptible AMS */
+		if (!atomic_test_bit(pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS)) {
```
This will always be true because of the check on line 248, right?

sambhurst replied:
I had the logic all wrong

DianaZig commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -264,13 +303,14 @@ void pe_report_error(const struct device *dev, const enum pe_error e,
 	}
+	/* Transmit error */
+	else if (e == ERR_XMIT) {
+		atomic_set_bit(pe->flags, PE_FLAGS_MSG_XMIT_ERROR);
```
Where does the soft reset occur for xmit errors now?  Just left to each PE state's discretion?

sambhurst replied:
Yes it's left up to the PE state's discretion.

DianaZig commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -597,24 +610,32 @@ static void prl_tx_wait_for_message_request_run(void *obj)
-			atomic_set_bit(&prl_tx->flags, PRL_FLAGS_WAIT_SINK_OK);
-			prl_tx_set_state(dev, PRL_TX_SNK_START_AMS);
+			if (pe_get_power_role(dev) == TC_ROLE_SOURCE) {
+				atomic_set_bit(&prl_tx->flags, PRL_FLAGS_SINK_NG);
```
Ah, perfect.  I'll mark this conversation as resolved as soon as I can find the button for it in GitHub...

DianaZig commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -713,6 +856,88 @@ static void pe_drs_send_swap_run(void *obj)
+	 * Shall send a Get_Sink_Cap Message and initialize and run the
+	 * SenderResponseTimer.
+	 */
+	pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_DR_SWAP);
```
Wrong control message?

sambhurst replied:
fixed

DianaZig commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -713,6 +856,88 @@ static void pe_drs_send_swap_run(void *obj)
+		if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+			pe->submachine = SM_WAIT_FOR_RX;
+		}
+		break;
```
nit: depending on whether you're using an external or internal TCPC, you might get TX complete notification at the same time as the message.  So in those cases, you can save yourself another state machine run by falling through

sambhurst replied:
done

DianaZig commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,691 @@
+	 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset state
+	 * when:
+	 *	1: There is an Explicit Contract and
+	 *	2: The Reject Message has been sent and the present Contract
```
For the requirement that the "reject message has been sent", do we need to check TX complete?

sambhurst replied:
I added the check

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,691 @@
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	LOG_INF("PE_SRC_Discover");
```
done

sambhurst commented on subsys/usb/usb_c/usbc_pe_src_states.c:
```diff
@@ -0,0 +1,691 @@
+		 * When message is sent, the Policy Engine Shall:
+		 *	1) Stop the NoResponseTimer .
+		 *	2) Reset the HardResetCounter and CapsCounter to zero.
+		 *	3) Initialize and run the SenderResponseTimer
```
I've added the code to stop the noResponse timer

keith-zephyr commented on doc/connectivity/usb_c/ucds.rst:
```diff
@@ -14,7 +14,7 @@ It provides the following functionalities:
   :zephyr_file:`include/zephyr/usb_c/usbc.h`
 
-Currently the device stack only support implementation of Sink devices.
+Currently the device stack supports implementation of Sink and Source devices.
```
```suggestion
Currently the device stack supports implementation of Sink only and Source only devices. Dual Role Power (DRP) devices are not yet supported.
```

sambhurst replied:
Done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -145,21 +157,190 @@ enum usbc_policy_wait_t {
+
 /** @cond INTERNAL_HIDDEN */
+/**
+ * @brief Callback type used to get the Sink Capabilities
+ *
+ * @param dev dev Runtime device structure
+ * @param pdos pointer where pdos are stored
+ * @param num_pdos pointer where number of pdos is stored
+ * @return 0 on success
```
Also, please put the src/snk/common callbacks into their own doxygen groups.

sambhurst replied:
Done

keith-zephyr replied:
You need to add this group to the API reference section of ucds.rst.

sambhurst committed:
usb-c: Refactor soft reset code

Refactor soft reset code

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Document the sink callbacks

Document the sink callbacks

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Properly detect reception of first AMS message

Detect and flag the reception of the first AMS mesage

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
usb-c: Add USB-C Subsystem Source PD Support

Add USB-C Power Delivery Source Support to USB-C
Subsystem

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -229,6 +234,39 @@ void pe_message_sent(const struct device *dev)
 	atomic_set_bit(pe->flags, PE_FLAGS_TX_COMPLETE);
 }
 
+static bool pe_soft_reset_is_required(const struct device *dev, const enum pd_packet_type type)
```
Done

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -264,13 +264,16 @@ void pe_report_error(const struct device *dev, const enum pe_error e,
 		pe_send_soft_reset(dev, type);
 	}
+	/* Transmit error */
+	else if (e == ERR_XMIT) {
```
Done

keith-zephyr replied:
You probably can just drop this commit, because you're refactoring and fixing the soft reset handling [here](https://github.com/zephyrproject-rtos/zephyr/pull/52993/commits/a616f11e92a0ef7f476f0c7af77f692f85d7562e#)

As is, this commit has an error regarding the non-AMS check - you also need to qualify that the message is an SOP.

sambhurst replied:
This commit adds the PE_FLAGS_MSG_XMIT_ERROR flag, so let's keep this commit instead of moving the flag to another commit.

alevkoy reviewed (commented):
I still have some comments open from my March 27 and 28 reviews. Mostly LGTM.

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -263,17 +296,15 @@ void pe_report_error(const struct device *dev, const enum pe_error e,
 	 *     Message is received instead of the expected GoodCRC Message
 	 *     response.
```
Move this comment block to the helper function.

sambhurst replied:
Done

keith-zephyr replied:
Note this is done with https://github.com/zephyrproject-rtos/zephyr/pull/52993/commits/cdc2e97f749fc203f231bc7db548a5aa0062ba52

keith-zephyr commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -229,6 +229,39 @@ void pe_message_sent(const struct device *dev)
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	/* Protocol Error not on SOP */
+	if (type != PD_PACKET_SOP) {
+		return false;
+	}
+
+	if (atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT)) {
+		/*
+		 * If the first Message in an AMS has been passed to the
+		 * Protocol Layer by the Policy Engine but has not yet been sent
+		 * when the Protocol Error occurs, the Policy Engine Shall Not
+		 * issue a Soft Reset
+		 */
+		if (!atomic_test_bit(pe->flags, PE_FLAGS_FIRST_MSG_SENT)) {
+			return false;
+		}
+
+		/*
+		 * If the Protocol Error occurs during an Interruptible AMS then
+		 * the Policy Engine Shall Not issue a Soft Reset
+		 */
+		if (atomic_test_bit(pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS)) {
+			return false;
+		}
+	}
```
It would be better to follow the language of the specification, even if this means the code is little bit larger.

```
if (SOP && !INTERRUPTIBLE_AMS)
    return true;

/* TODO - Check if a message has not been sent after retries to the Sink */

if (!EXPLICIT_CONTRACT) {
    if (SOP && INTERRUPTIBLE_AMS)
        return true;
    if (SOP && !FIRST_MESSAGE_SENT)
        return true; 
}
return false;
```

sambhurst replied:
From our offline converstion, the spec describes both implementations.

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -146,20 +146,80 @@ enum usbc_policy_wait_t {
+/**
+ * @brief Callback type used to get the Sink Capabilities
+ *
+ * @param dev dev Runtime device structure
```
Please specify the type of device passed in for all these prototypes.
```suggestion
 * @param dev USB-C Connector Instance
```

sambhurst replied:
Done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -146,20 +146,80 @@ enum usbc_policy_wait_t {
 				       uint32_t **pdos,
 				       int *num_pdos);
+/**
+ * @brief Callback type used to store the received Port Partner's
+ *	  Source Capabilities
```
```suggestion
 * @brief Callback type used to report the received Port Partner's
 *	  Source Capabilities
```

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -146,20 +146,80 @@ enum usbc_policy_wait_t {
+ *	  Source Capabilities
+ *
+ * @param dev dev Runtime device structure
+ * @param pdos pointer where pdos are stored
+ * @param num_pdos number of pdos stored
```
```suggestion
 * @param pdos pointer to the partner's source pdos
 * @param num_pdos number of source pdos
 * ```

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -146,20 +146,80 @@ enum usbc_policy_wait_t {
+ *
+ * @param dev dev Runtime device structure
+ * @param policy_check policy to check
+ * @return true if check passes
```
```suggestion
 * @return true if policy is currently allowed by the device policy manager
```

keith-zephyr commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -47,6 +47,8 @@ enum prl_flags {
 	 */
 	PRL_FLAGS_MSG_XMIT = 7,
+	/** Flag to track if first message in AMS is pending */
+	PRL_FLAGS_FIRST_MSG_PENDING = 8,
```
Commit [usb-c: Properly detect reception of first AMS message](https://github.com/zephyrproject-rtos/zephyr/pull/52993/commits/1e9fb5a23ad27ca86715b9bb79c4ec577ebe23bd) commit needs to be the parent of the [usb-c: Refactor soft reset code](https://github.com/zephyrproject-rtos/zephyr/pull/52993/commits/a616f11e92a0ef7f476f0c7af77f692f85d7562e)

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +244,109 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
  */
 typedef bool (*policy_cb_is_snk_at_default_t)(const struct device *dev);
 /** @}*/
+/** \addtogroup source_callbacks
```
Add this doxygen group to the API reference of ucds.rst.

Also update the `dev` parameter documentation to `USB-C Connector Instance`

sambhurst replied:
Should the typedef be listed of the ucds.rst api, because they're in /** @cond INTERNAL_HIDDEN */ section of the include.

keith-zephyr replied:
These callbacks are part of the public API.  The should be moved out if the hidden section.

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +244,109 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+ * @param request_msg request message to check
+ * @return sink request reply
+ */
+typedef enum usbc_snk_req_reply_t (*policy_cb_check_sink_request_t)(
+				const struct device *dev,
+				const uint32_t request_msg);
```
Looks like this file needs clang-format.

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +244,109 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+ * @param dev dev Runtime device structure
+ * @return true if power supply is ready, else false
+ */
+typedef bool (*policy_cb_check_if_ps_ready_t)(const struct device *dev);
```
Nit - you could name this `policy_cb_is_ps_ready_t`

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +244,109 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+ * @param present_contract present contract
+ * @return true if present contract is still valid
+ */
+typedef bool (*policy_cb_check_present_contract_t)(const struct device *dev,
```
Nit - `policy_cb_present_contract_is_valid`

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +244,109 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+typedef bool (*policy_cb_change_src_caps_t)(const struct device *dev);
+
+/**
+ * @brief Callback type used to store the Capabilities received from
```
```suggestion
 * @brief Callback type used to report the Capabilities received from
```

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -146,20 +146,80 @@ enum usbc_policy_wait_t {
+/**
+ * @brief Callback type used to check a policy
+ *
+ * @param dev dev Runtime device structure
```
```suggestion
 * @param dev USB-C Connector Instance
```

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,6 +241,105 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+ * @brief Callback type used to get the Rp value that should be placed on
+ *        the CC lines
+ *
+ * @param dev dev Runtime device structure
```
```suggestion
 * @param dev USB-C Connector Instance
```

Replace all instances for the APIs your are introducing in this commit.

sambhurst replied:
done

keith-zephyr commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -220,7 +240,105 @@ typedef uint32_t (*policy_cb_get_rdo_t)(const struct device *dev);
+ *        the CC lines
+ *
+ * @param dev USB-C Connector Instance
+ * @param rp
```
Missing description for the `rp` parameter.  I thought the CI had checks for this.

sambhurst replied:
done

keith-zephyr reviewed (commented):
One last nit on the documentation from me.

kartben commented on include/zephyr/usb_c/usbc.h:
```diff
@@ -145,22 +157,188 @@ enum usbc_policy_wait_t {
+ *  @{
+ */
+/**
+ * @brief Callback type used to get the Rp value that should be placed on
+ *        the CC lines
+ *
+ * @param dev USB-C Connector Instance
+ * @param rp Rp value
+ * @return 0 on success
+ */
+typedef int (*policy_cb_get_src_rp_t)(const struct device *dev, enum tc_rp_value *rp);
+
+/**
+ * @brief Callback type used to enable VBUS
+ *
+ * @param dev USB-C Connector Instance
+ * @param en true to enable VBUS, else disable it
+ * @return 0 on success
+ */
+typedef int (*policy_cb_src_en_t)(const struct device *dev, bool en);
```
https://github.com/zephyrproject-rtos/zephyr/pull/57928
